xunit-autolabeler-v2/ast_parser/python/test_parser.py

```diff
@@ -0,0 +1,227 @@
+from . import drift_test
+
+
+def _get_test_nodes(parsed_nodes: List[Any]) -> List[Any]:
+    """Helper function to extract test methods wrapped in classes
+
+    Args:
+        parsed_nodes (List[ast.AST]): a list of nodes parsed
+                                      from a Python test file
+
+    Returns:
+        List[ast.AST]: a list of possible test nodes
+    """
+
+    possible_test_nodes = []
+    for node in parsed_nodes:
+        possible_test_nodes.append(node)
+        if 'ClassDef' in str(type(node)):
+            # Concatenate node.body (which is an array)
+            possible_test_nodes += node.body
+
+    return [node for node in possible_test_nodes
+            if hasattr(node, 'name') and node.name.startswith('test_')]
+
+
+def get_test_methods(test_path: str) -> List[Any]:
+    """Gets the top-level methods within a test file
+
+    Args:
+        source_path: path to the file to process
+
+    Returns:
+        List[ast.AST]: a list of the top-level
+                       methods within the provided file
+    """
+    try:
+        with open(test_path, 'r') as file:
+            content = "".join(file.readlines())
+            parsed_nodes = list(ast.iter_child_nodes(ast.parse(content)))
+
+            test_nodes = _get_test_nodes(parsed_nodes)
+
+            for node in test_nodes:
+                node.test_path = test_path
+
+            # Verify file contains no duplicate method names
+            # (Only relevant for test methods wrapped in classes)
+            used_test_names = set()
+            for node in test_nodes:
+                if node.name in used_test_names:
+                    raise ValueError(
+                        f'Test name {node.name} in file'
+                        f' {test_path} must be unique.')
+                used_test_names.add(node.name)
+
+            return test_nodes
+    except IOError as err:
+        # Fail gracefully if a file can't be read
+        # (This shouldn't happen, but if it doess
+        #  we don't want to 'break the build'.)
+        sys.stderr.write(
+            f'WARNING: could not read file: {test_path}\n')
+        sys.stderr.write(
+            f'\t{str(err)}\n')
+
+        return []
+
+
+def get_test_key_to_snippet_map(
+    test_methods: List[Any]
+) -> Dict[Tuple[str, str], Tuple[str, str]]:
+    """Map the supplied test methods to their relevant 'test keys'
+
+    Test keys are tuples that a) identify a particular snippet and b)
+    can be matched with that snippet's tests. The attributes used in
+    a test key differ based on which parser detected a given snippet.
+
+    Args:
+        test_methods (List[ast.AST]): a list of methods extracted
+                                      from test files
+
+    Returns: a mapping between test keys and their corresponding test data
+    """
+    test_to_method_key_map = {}
+
+    def __recursor__(expr: Any) -> List[drift_test.DriftTest]:
+        """Recursively find test keys within an expression.
```

Comment:
Small nit:

I'd make the docstring style consistent. Here you have a period at the end of the first line, whereas you don't have period in other places.

Reply:
Done.